### ISSUE TITLE
SDXL: Inconsistent VRAM allocation with sample enabled, but stable without

### ISSUE THREAD

**Issue by SharkWipf** [kohya-ss/sd-scripts#2030]
As per the title:
![Image](https://github.com/user-attachments/assets/c516a46d-476c-43c2-b523-8ca938e6b338)

With sample image generation enabled, memory allocation seems to jump up and down every time a sample has been generated.
Without, it stays consistent throughout the entire run, maybe creeping up a ~200MB across the whole run at most.
The same run, with the same settings, eventually ran out of VRAM with sampling enabled, and completed fine without.

I don't have the exact config anymore, but it was close to my current one, but with sampling enabled, a sampling seed configured and the sampler set to Euler Ancestral I think:

```bash
accelerate launch sdxl_train.py \
 --dataset_config ../config/sdxl_finetune.toml \
 --output_dir "/home/sebastiaan/ml/dan/training/${output_name}" \
 --output_name "${output_name}" \
 --save_model_as safetensors \
 --pretrained_model_name_or_path "/home/sebastiaan/ml/dan/training/out_sd-scripts_sdxl_obs_01/sd-scripts_sdxl_obs_01_diffusers/" \
 --vae "madebyollin/sdxl-vae-fp16-fix" \
 --seed 42 \
 --train_batch_size 8 \
 --learning_rate 6e-6 \
 --mixed_precision bf16 \
 --save_precision bf16 \
 --gradient_accumulation_steps 4 \
 --gradient_checkpointing \
 --train_text_encoder \
 --learning_rate_te1 2e-6 \
 --learning_rate_te2 1e-6 \
 --save_every_n_steps 250 \
 --save_last_n_steps 250 \
 --deepspeed \
 --sdpa \
 --torch_compile \
 --mem_eff_attn \
 --dynamo_backend inductor \
 \
 --max_train_epochs 1 \
 --log_with wandb \
 --log_tracker_name "${output_name}" \
 --wandb_run_name "${output_name}" \
 --optimizer_type AdamW \
 --lr_scheduler polynomial \
 --lr_warmup_steps 0 \
 --max_token_length 225
```

Using DeepSeek stage2 with CPU offloading.

**Comment by SharkWipf** [comment]
Hmm, perhaps it is not properly clearing the CUDA cache after it's done generating and unloading the sample image(s)? Because VRAM seems to drop correctly when generating the image, it's only after it resumes training that memory increases again. 
I'll test more later, I currently have 24 hours left to go on a training run that I don't want to interrupt.

**Comment by SharkWipf** [comment]
Like #2029, this issue most likely stems from Deepspeed estimating allocation VRAM wrong.
Can it be fixed? Yeah, probably, but it's probably a lot of hassle and not worth the effort.
Since I can't run without Deepspeed to test unless I'm training a LoRA, and since I am not training any LoRAs atm, I'll close this issue now. If I run into it without Deepspeed too I'll reopen it again.
